File: Data Structures & Algorithms/three-integer-sum/submission-26.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = set()
        nums.sort()

        for i in range(len(nums)):
            complement = -nums[i]

            l = i + 1
            r = len(nums) - 1

            while l < r:
                if nums[l] + nums[r] == complement:
                    res.add(tuple([nums[i], nums[l], nums[r]]))
                    # found a valid triplet search for more
                    l += 1
                    r -= 1
                elif nums[l] + nums[r] < complement:
                    l += 1
                else:
                    r -= 1

        return [list(x) for x in res]

File: Data Structures & Algorithms/house-robber/submission-26.java
class Solution {
    public int rob(int[] nums) {
        int n = nums.length;

        if (n == 0) {
            return 0;
        }
        if (n == 1) {
            return nums[0];
        }
        
        int[] dp = new int[n];

        dp[n - 1] = nums[n - 1];
        dp[n - 2] = Math.max(nums[n - 2], nums[n - 1]);

        for (int i = n - 3; i >= 0; i--) {
            dp[i] = Math.max(nums[i] + dp[i + 2], dp[i + 1]);
        }

        return dp[0];
    }
}
